app/models/queries/projects/selects/life_cycle_step.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+# -- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+# ++
+
+class Queries::Projects::Selects::LifeCycleStep < Queries::Selects::Base
+  KEY = /\Alcsd_(\d+)\z/
+
+  def self.key
+    KEY
+  end
+
+  def self.all_available
+    return [] unless available?
+
+    Project::LifeCycleStepDefinition
+      .pluck(:id)
+      .map { |id| new(:"lcsd_#{id}") }
+  end
+
+  def caption
+    life_cycle.name
+  end
+
+  def life_cycle
+    return @life_cycle if defined?(@life_cycle)
+
+    @life_cycle = Project::LifeCycleStepDefinition
+                    .find_by(id: attribute[KEY, 1])
+  end
+
+  def available?
+    life_cycle.present?
+  end
+
+  def action_menu_header(button)
+    # Show the proper icon for the definition with the correct color.
+    icon = case life_cycle
+           when Project::StageDefinition
+             :"git-commit"
+           when Project::GateDefinition
+             :diamond
+           else
+             raise "Unknown life cycle definition for: #{life_cycle}"
+           end
+
+    classes = helpers.hl_inline_class("life_cycle_step_definition", life_cycle)
+    button.with_leading_visual_icon(icon:, classes:)
+
+    # As all other action menu headers, we will show an action icon and the caption:
+    button.with_trailing_action_icon(icon: :"triangle-down")
+
+    caption.to_s
+  end
```

Comment:
Just a preference, but I would rather keep the button and caption in the sort_helper and just return the `{ icon:, classes: }` from this method. The [sort_helper](https://github.com/opf/openproject/pull/17400#discussion_r1882735110) would also get simpler.

```suggestion
  def visual_icon(button)
    # Show the proper icon for the definition with the correct color.
    icon = case life_cycle
           when Project::StageDefinition
             :"git-commit"
           when Project::GateDefinition
             :diamond
           else
             raise "Unknown life cycle definition for: #{life_cycle}"
           end

    classes = helpers.hl_inline_class("life_cycle_step_definition", life_cycle)
    { icon:, classes: }
  end
```

Reply:
That's neat! It bothered me that the Select got too involved in the UI. It's much cleaner like this. Nice!